### PR TITLE
Migrate Chart component to X charts library

### PR DESCRIPTION
**Pull request by apedroferreira**
Use `@mui/x-charts` for the Chart component. 
Includes some general improvements/fixes for charts too.

Closes https://github.com/mui/mui-toolpad/issues/2309
Also closes https://github.com/mui/mui-toolpad/issues/2542
Also closes https://github.com/mui/mui-toolpad/issues/2592
Also closes https://github.com/mui/mui-toolpad/issues/2602

Screenshot 2023-10-31 at 19 58 05

### CODE REVIEW

**Comment by apedroferreira** [comment]
Still running into issues with imports: the build output chunks from `toolpad-components` seem not to able to resolve the import from `@mui/x-charts/ChartContainer`, which is in the root `node_modules` folder...

Any ideas @Janpot ? I've tried a few things but no success so far, let me know if you can help figure this one out once you're back!

Screenshot 2023-08-16 at 18 21 52

**Comment by Janpot** [comment]
cc @alexfauquette 

Looks like I can fix this by changing the `import` condition of the `./*` export in the `@mui/x-charts` `package.json` file to `"./esm/*/index.js"`
```json
    "./*": {
      "types": "./*/index.d.ts",
      "require": "./*",
      "import": "./esm/*/index.js"
    }
```

**Comment by alexfauquette** [comment]
> Looks like I can fix this by changing the import condition of the ./* export in the @mui/x-charts package.json file to "./esm/*/index.js"

Should be fixed by the following PR

https://github.com/mui/mui-x/pull/10111

**Comment by alexfauquette** [comment]
I don't know why Github closed this PR

**Comment by apedroferreira** [comment]
> I don't know why Github closed this PR

No worries :D Thanks a lot Alex!

**Comment by apedroferreira** [comment]
Created an issue with feedback/issues I ran into in https://github.com/mui/mui-x/issues/10140.

File: packages/toolpad-app/src/server/toolpadAppBuilder.ts [hunk 1]
```diff
@@ -297,7 +306,12 @@ export function createViteConfig({
       appType: 'custom',
       logLevel: 'info',
       root,
-      plugins: [react(), toolpadVitePlugin({ root, base, getComponents }), ...plugins],
+      plugins: [
+        react(),
+        viteCommonjs(),
```
Comment (oliviertassinari) [inline packages/toolpad-app/src/server/toolpadAppBuilder.ts:311]: ?
```suggestion
        // TODO, remove once https://github.com/mui/mui-x/pull/9826 is fixed
        viteCommonjs(),
```

File: packages/toolpad-app/jest.config.js [hunk 2]
```diff
@@ -7,6 +7,18 @@ const transpiledModules = [
...
+  'd3-array',
+  'internmap',
+  'd3-interpolate',
+  'd3-color',
+  'd3-format',
+  'd3-time',
+  'd3-time-format',
+  'd3-shape',
+  'd3-path',
+  '@mui/x-charts/esm',
```
Comment (Janpot) [inline packages/toolpad-app/jest.config.js:20]: that path is not exported from `@mui/x-charts`
```suggestion
  '@mui/x-charts',
```

Comment (apedroferreira) [reply packages/toolpad-app/jest.config.js:20]: `@mui/x-charts` wasn't working to stop jest from showing an error, but `@mui/x-charts/esm` did... Maybe it's possible to use a glob or something though, I can try that.

Comment (oliviertassinari) [reply packages/toolpad-app/jest.config.js:20]: We don't allow imports from the esm folder: https://mui.com/material-ui/guides/minimizing-bundle-size/#how-to-use-custom-bundles.

Comment (Janpot) [reply packages/toolpad-app/jest.config.js:20]: @apedroferreira just `'@mui/x-charts',` seems to work for me.

I'm going to try a migration to `vitest`, I've just about had it hunting down ESM issues with vite.

Comment (Janpot) [reply packages/toolpad-app/jest.config.js:20]: => https://github.com/mui/mui-toolpad/pull/2709

This took a fraction of the time I've spent fighting jest about ESM imports. Had to configure nothing at all, just works out of the box.

**Review by bharatkashyap — APPROVED** [review]
Looks good! Are there any documentation updates which would need to be made?

**Comment by apedroferreira** [comment]
> Looks good! Are there any documentation updates which would need to be made?

We don't have very deep documentation for the Chart, just the automatic one... 
I guess I could add something, probably in a separate PR.

**Review by Janpot — APPROVED** [review]
Nice!

**Comment by apedroferreira** [comment]
Merging, I'll keep it noted to add documentation for the Chart component.

**Pull request merged**